capstone/src/main/webapp/js/autocomplete.js

```diff
@@ -0,0 +1,169 @@
+ */
+function addActiveTag(autocompleteListElmt) {
+  if (!autocompleteListElmt) {
+    return;
+  }
+  removeActiveTag(autocompleteListElmt);
+  if (selectedElmntPos >= autocompleteListElmt.length) {
+    selectedElmntPos = 0;
+  } else if (selectedElmntPos < 0) {
+    selectedElmntPos = (autocompleteListElmt.length - 1);
+  }
+  autocompleteListElmt[selectedElmntPos].classList.add(ACTIVE_CLASS);
+}
+
+/**
+ * Removes the 'active' tag from an autocomplete elmt
+ * @param {Object} autocompleteListElmt
+ */
+function removeActiveTag(autocompleteListElmt) {
+  for (elmt of autocompleteListElmt) {
+    elmt.classList.remove(ACTIVE_CLASS);
+  }
+}
+
+/**
+ * Closes dropdown autocomplete list
+ * Does not close the element provided
+ * @param {Object} elmnt
+ */
+function closeAllElmntExcept(elmnt) {
+  const autoItems = document.getElementsByClassName(ITEM_CLASS);
+  for (const item of autoItems) {
+    if (item != elmnt && item != SEARCH_INPUT) {
+      item.parentNode.removeChild(item);
+    }
+  }
+}
+
+/**
+ * Adds valid suggestions to list of stored suggestions
+ * @param{string} validSuggestion
+ */
+function keepTrackOfSuggestions(validSuggestion) {
+  const parent = document.getElementById('suggestions');
+  const listElt = document.createElement('li');
+  listElt.innerHTML = validSuggestion;
+  parent.appendChild(listElt);
+}
+
+/**
+ * Current placeholder until servlet is created
+ * @return{Array<string>} arr
+ */
+function getListOfSuggestions() {
+  const arr = ['Hallo', 'Hello', 'Hi', 'Hiya', 'Howdy', 'Wassup'];
+  return arr;
+}
+
+/**
+ * Identifies and returns relevant suggestions in the arr
+ * @param {string[]} arr
+ * @param {string} val
+ * @return {string[]} trimArr
+ */
+function getRelevantDataSuggestions(arr, val) {
+  const trimArr = [];
+  val = val.toUpperCase();
+  for (const arrElt of arr) {
+    if (arrElt.substr(0, val.length).toUpperCase() === val.toUpperCase()) {
+      trimArr.push(arrElt);
+    }
+  }
+  return trimArr;
+}
+
+/**
+ * Takes relevant suggestions and displays them in DOM
+ * @param {Object} trimArr
+ * @param {Object} autocompleteList
+ * @param {string} val
+ */
+function displaySuggestions(trimArr, autocompleteList, val) {
+  for (arrElt of trimArr) {
+    const listElmt = document.createElement('div');
+    listElmt.innerHTML = '<strong>' +
+                         arrElt.substr(0, val.length) +
+                         '</strong>';
+    listElmt.innerHTML += arrElt.substr(val.length);
+    const cpyArrElt = arrElt;
+    listElmt.addEventListener('click', () => {
+      SEARCH_INPUT.value = cpyArrElt;
+      keepTrackOfSuggestions(cpyArrElt);
+      closeAllElmntExcept();
+    });
+    autocompleteList.appendChild(listElmt);
+  }
+}
+
+/**
+ * Event occurrance when input is provided to autocomplete field
+ */
+function giveSuggestions() {
+  closeAllElmntExcept();
+  const val = SEARCH_INPUT.value;
+  if (!val) {
+    return;
+  }
+  if (SEARCH_INPUT.value !== val) {
+    return;
+  }
```

Comment:
What's the purpose of this condition? To the casual reader it appears that it would never return since val was just assigned to SEARCH_INPUT.value.

Reply:
Kaveh suggested addition in an earlier comment. It did not break anything so I decided to keep it in. I can remove though.

Reply:
My suggestion was to return early without updating the list element in order to prevent stale data from being shown to user. This was meant to be a basic implementation to prevent race conditions. So we do `val = SEARCH_INPUT.value` before fetch and once the fetch completed, we add this if statement. Not sure why i didn't work. But in any case, feel free to remove it and ignore my solution if it's not working.